Add option to load module before anything else #314

anthonyringoet opened the pull request:
As described in #314.

Some notes:
- `-r` option was already in use for `prefix` so took `b`/`before-module` (like existing `logging-module` option). Feel free to request other naming if preferred.
- Requiring the actual module seemed exactly the same as the existing `logging-module`. In an effort to create a function out of it, I abstracted the core part of `util.js's` `requireServerPluginFromPath` to `requireModule`. This leads to some minor differences in textual error output. Behavior of when errors are thrown is still the same though.


<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

mcollina commented:
I think we should keep the same option name of Node.js, following the principle of least surprise (we will bump the major).

mcollina commented on start.js:
```diff
@@ -59,8 +59,16 @@ function stop (message) {
-
   const opts = parseArgs(args)
+
+  if (opts.beforeModule) {
```
how would this be handled for the `watch` option?

anthonyringoet replied:
Watch option checks for changes in `process.cwd()` so currently for both this addition and the custom logger module it won't be triggered if it's outside of the current working directory. Should both of these files be watched explicitly as well?

mcollina reviewed (changes requested):
This _might_ not be exactly what @delvedor needs. If I understand correctly, he needs this to load Elastic APM before everything else. 

Specifically, this should be loading only commonjs and not ESM, and possibly done earlier in the process startup (right after argument parsing.

anthonyringoet commented:
Makes sense and will make it a lot more straightforward if it can move a level up towards the `start` function. Will look into these changes.

mcollina reviewed (approved):
lgtm

climba03003 reviewed (commented):
I think the test case about `require` and `fail to require` is missing. 
1. `require` is really done before the `fastify` is running
2. the behavior when `require is fail to load`

The first test is important as it is the actual function is this PR.

zekth reviewed (changes requested):
Like @climba03003 said, missing the tests cases of require working and not working.

Otherwise lgtm

anthonyringoet commented:
Added support for multiple `-r` flags (same as node cli) and added tests.

@climba03003 How would you go about testing if the require(s) are really done before the fastify server is running?

climba03003 commented:
> 
> 
> Added support for multiple `-r` args (same as node cli) and added tests.
> 
> @climba03003 How would you go about testing if the require(s) are really done before the fastify server is running?

Just like your other test case, you can mutate global variables. Then, you can execute `tap` inside the server files.
It can ensure you receive a the things you want in the server.

I think we can expect, if `required files` use `top-level async` method. The order will break? should this be tested or mention?

anthonyringoet commented:
@climba03003 Like the previous commit?

The top-level async behavior is the same with `node -r ./some-async-file.js` afaik.

climba03003 reviewed (approved):
Good to know that it is the same as `node -r`.

mcollina commented:
@zekth is this good to land now?

zekth reviewed (approved):
good work!
